src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs

```diff
@@ -136,34 +142,80 @@ public static PropertyHelper[] GetVisibleProperties(Type type)
             // Instance methods in the CLR can be turned into static methods where the first parameter
             // is open over "target". This parameter is always passed by reference, so we have a code
             // path for value types and a code path for reference types.
-            var typeInput = getMethod.DeclaringType;
-            var typeOutput = getMethod.ReturnType;
+            if (getMethod.DeclaringType.GetTypeInfo().IsValueType)
+            {
+                // Create a delegate (ref TDeclaringType) -> TValue
+                return MakeFastPropertyGetter(
+                    typeof(ByRefFunc<,>),
+                    getMethod,
+                    CallPropertyGetterByReferenceOpenGenericMethod);
+            }
+            else
+            {
+                // Create a delegate TDeclaringType -> TValue
+                return MakeFastPropertyGetter(
+                    typeof(Func<,>),
+                    getMethod,
+                    CallPropertyGetterOpenGenericMethod);
+            }
+        }
+
+        /// <summary>
+        /// Creates a single fast property getter which is safe for a null input object. The result is not cached.
+        /// </summary>
+        /// <param name="propertyInfo">propertyInfo to extract the getter for.</param>
+        /// <returns>a fast getter.</returns>
+        /// <remarks>
+        /// This method is more memory efficient than a dynamically compiled lambda, and about the
+        /// same speed.
+        /// </remarks>
+        public static Func<object, object> MakeNullSafeFastPropertyGetter(PropertyInfo propertyInfo)
```

Comment:
this and `MakeFastPropertyGetter()` are nearly identical.  suggest refactoring to avoid the duplication e.g. pass in the `callValueGetter` and `callNonValueGetter` strings to a new method.